x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/TopNOperator.java

```diff
@@ -22,12 +24,47 @@
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.BitSet;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
+import java.util.function.BiConsumer;
 
 public class TopNOperator implements Operator {
 
+    private static final byte SEPARATOR = 0x0; // separator for values inside the BytesRef sorting key
+    private static final byte SMALL_NULL = 0x01; // "null" representation for "nulls first"
+    private static final byte BIG_NULL = 0x02; // "null" representation for "nulls last"
+
+    public enum Encoder {
+        IP_BYTES_REF_ENCODER((rowValue, bytesRefBuilder) -> bytesRefBuilder.append(rowValue)),
+        BYTES_REF_ENCODER((rowValue, bytesRefBuilder) -> {
+            // add one bit to every byte so that there are no "0" bytes in the provided bytes. The only "0" bytes are
+            // those defined as separators
+            int end = rowValue.offset + rowValue.length;
+            for (int i = rowValue.offset; i < end; i++) {
+                byte b = rowValue.bytes[i];
+                if ((b & 0b1000_0000) == 0) {
+                    b++;
```

Comment:
Not sure this works as it is:  with this encoding, `0b0111_1111` and `0b1000_0000` become the same value, so you lose the ordering. 
The idea seems promising though, instead of setting one bit to 1, you have to _add_ one more bit for every 7 bits in the input. The algorithm is more convoluted and it will consume more heap memory, but you'll also save some space from the null bits (you need only two - maybe three - bits, not a full byte)

Reply:
Indeed, by incrementing 0x7F it will become 0x80, which is valid as last byte in any non-ascii UTF-8 character. I think mangling just the ascii bytes won't work.
There are a few bytes which cannot appear in a valid UTF-8 sequence and safest would be 0xFE or 0xFF, though these wouldn't have the nice sorting property of 0.
But this also means that incrementing _all_ bytes would be possible, though not sure if this would be more performant than not using a "composite" key.

Reply:
I remember running through this algorithm and thinking "this works". I can't reproduce it in my head now, but I was convinced at some point. I recall it had something to do with the utf-8 encoding itself - like, `0b0111_1111` can't appear in a position where `0b1000_0000` appears. The valid high nibbles in start-of-character bytes are `0xxx`, `110x`, `1110`, and `1111` and `10xx` is only available for continuations.

Reply:
> I recall it had something to do with the utf-8 encoding itself - like, 0b0111_1111 can't appear in a position where 0b1000_0000 appears

I guess it applies to strings only. My concern is on generic types and bytes representations, since `public BytesRef(byte[] bytes)` allows to pass any sequence (eg. [binary](https://www.elastic.co/guide/en/elasticsearch/reference/current/binary.html) type), and eventually we'll support most of them.

Reply:
For `binary` that could be a problem, yes, BUT is it feasible/realistic to sort on a binary field? I've talked to Nik some time ago about this and there is a real possibility to not allow sorting on `binary` fields in ESQL by design.

@luigidellaquila @bpintea you both bring good points about sorting, but I argue these are pure theoretical that haven't been verified in a **failed unit test**. For me this matters a great deal, because I believe the tests are random enough and cover a wide range on values, and I've run hundreds of thousand of tests (even more than a million) until this point. So, either the tests are faulty or your theory doesn't stand.

My ask for you is to:
- find a flaw in the current unit tests here https://github.com/elastic/elasticsearch/pull/98337/files#diff-1959fc38423598cf6033ebf73609a035e5ad5c47fbfe8b21ea586eed526becb1R830
- and/or (preferably "and") come up with a realistic standalone unit test proving your theories, similar to https://github.com/elastic/elasticsearch/pull/98337/files#diff-1959fc38423598cf6033ebf73609a035e5ad5c47fbfe8b21ea586eed526becb1R1125

Reply:
I think the tests are good, the problem is that they only cover currently supported field types (ie. practically only strings), but the code seems to assume that BYTES_REF_ENCODER covers all the possible BytesRef representations, blacklisting IP as a special case (while, in a more general sense, the special case here is strings).
Unfortunately it could not be the case: from Nik's comment `I recall it had something to do with the utf-8 encoding itself - like, 0b0111_1111 can't appear in a position where 0b1000_0000 appears`, I doubt it also applies to Version, Point, Shape just to name a few. 
I'd be good with this solution as long as we change BYTES_REF_ENCODER to STRING_ENCODER and we throw some kind of exception for other types, so that when we add a new type we don't silently introduce bugs.

Reply:
> `0b0111_1111` can't appear in a position where `0b1000_0000` appears.

Good point: before `0b1000_0000` there'd have to be a `1xxx xxx` byte (either start-of-character or continuation), but which can't appear before `0b0111_1111`. Meaning my UTF-8 concern above isn't actually justified.

Reply:
>  I doubt it also applies to Version, Point, Shape just to name a few

Good point about these. I know that `version` has a special encoder and it's a scenario that needs further investigation and, off the top of my head, is a good candidate for not using the BytesRef encoder on it (just like for IPs). For the other types.... I have no knowledge on how they are encoded/sorted on, but coming up with a mechanism as a "heads up" warning for new supported types when it comes to sorting in TopN is a good idea.

Reply:
A trivial test case is the following:
```
       public void testCompareBytesRef() {
        Block[] bs = new Block[] {
            BytesRefBlock.newBlockBuilder(2)
                .appendBytesRef(new BytesRef(new byte[] { 127 }))
                .appendBytesRef(new BytesRef(new byte[] { -128 }))
                .build() };
        Page page = new Page(bs);
        TopNOperator.RowFactory rowFactory = new TopNOperator.RowFactory(page);

        Tuple<TopNOperator.Row, TopNOperator.Row> rows = bytesRefRows(
            false,
            randomBoolean(),
            so -> rowFactory.row(page, 0, null, so),
            null,
            0
        );

        rows = bytesRefRows(true, randomBoolean(), so -> rowFactory.row(page, 0, null, so), so -> rowFactory.row(page, 1, null, so), 0);
        assertThat(TopNOperator.compareRows(rows.v1(), rows.v2()), greaterThan(0));
    }
```

I don't have a complete knowledge of the encoding of all types supported by ES, so maybe I'm just overestimating the impact of this thing.
Up to you folks, the rest looks good